astropy/cosmology/io/model.py

```diff
@@ -0,0 +1,262 @@
+
+    See Also
+    --------
+    astropy.cosmology.Cosmology.to_format
+    """
+
+    @abc.abstractmethod
+    def _cosmology_class(self):
+        """Cosmology class as a private attribute. Set in subclasses."""
+
+    @abc.abstractmethod
+    def _method_name(self):
+        """Cosmology method name as a private attribute. Set in subclasses."""
+
+    @classproperty
+    def cosmology_class(cls):
+        """|Cosmology| class."""
+        return cls._cosmology_class
+
+    @property
+    def cosmology(self):
+        """Return |Cosmology| using `~astropy.modeling.Parameter` values."""
+        cosmo = self._cosmology_class(
+            name=self.name,
+            **{k: (v.value if not (v := getattr(self, k)).unit else v.quantity)
+               for k in self.param_names})
+        return cosmo
+
+    @classproperty
+    def method_name(self):
+        """Redshift-method name on |Cosmology| instance."""
+        return self._method_name
+
+    # ---------------------------------------------------------------
+
+    def evaluate(self, *args, **kwargs):
+        """Evaluate method {method!r} of {cosmo_cls!r} Cosmology.
+
+        The Model wraps the :class:`~astropy.cosmology.Cosmology` method,
+        converting each |Cosmology| :class:`~astropy.cosmology.Parameter` to a
+        :class:`astropy.modeling.Model` :class:`~astropy.modeling.Parameter`
+        (unless the Parameter is None, in which case it is skipped).
+        Here an instance of the cosmology is created using the current
+        Parameter values and the method is evaluated given the input.
+
+        Parameters
+        ----------
+        *args, **kwargs
+            The first ``n_inputs`` of ``*args`` are for evaluating the method
+            of the cosmology. The remaining args and kwargs are passed to the
+            cosmology class constructor.
+            Any unspecified Cosmology Parameter use the current value of the
+            corresponding Model Parameter.
+
+        Returns
+        -------
+        Any
+            Results of evaluating the Cosmology method.
+        """
+        # create BoundArgument with all available inputs beyond the Parameters,
+        # which will be filled in next
+        ba = self.cosmology_class._init_signature.bind_partial(*args[self.n_inputs:], **kwargs)
+
+        # fill in missing Parameters
+        for k in self.param_names:
+            if k not in ba.arguments:
+                v = getattr(self, k)
+                ba.arguments[k] = v.value if not v.unit else v.quantity
+
+            # unvectorize, since Cosmology is not vectorized
+            # TODO! remove when vectorized
+            if np.shape(ba.arguments[k]):  # only in __call__
+                # m_nu is a special case  # TODO! fix by making it 'structured'
+                if k == "m_nu" and len(ba.arguments[k].shape) == 1:
+                    continue
+                ba.arguments[k] = ba.arguments[k][0]
```

Comment:
Ah. So the arguments to most Cosmology methods can be scalars or arrays, but the parameters used to initialize a Cosmology must be scalars, scalar-like (e.g. scalar Quantity), or `None`.  This argument hack is just to pop a parameter back to a scalar, because Model.__call__ always hands over ``[parameter]``, even if ``parameter`` is a scalar.
This doesn't influence the underlying function which can accept a multidimensional input array.

e.g.
```python
cosmo = Cosmology(**<parameters must be scalar>)
cosmo.method(*<arguments can be ND arrays>)
```